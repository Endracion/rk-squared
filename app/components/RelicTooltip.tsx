import * as React from 'react';
import * as ReactTooltip from 'react-tooltip';

import enlir from '../data/enlir';

import * as _ from 'lodash';

interface Props {
  id: string;
  relicId: number;
}

const attributes = ['Attack', 'Defense', 'Magic', 'Resistance', 'Mind', 'Accuracy', 'Evasion'];

const separateWithBr = (lines: any[]) => {
  const result: any[] = [];
  for (const i of lines) {
    if (result.length) {
      result.push(<br/>);
    }
    result.push(i);
  }
  return result;
};

export class RelicTooltip extends React.Component<Props & any> {
  render() {
    const { id, relicId, ...props } = this.props;
    if (!enlir.relics[relicId]) {
      return null;
    }
    const relic = enlir.relics[relicId];

    const lines = _.filter([
      `${relic.Rarity}â˜… ${enlir.types.relics[relic.RelicType]}`,
      _.filter(attributes.map(i => relic[i] ? `${i} ${relic[i]}` : '')).join(', '),
      relic.Effect
    ]);

    return (
      <ReactTooltip id={id} {...props}>
        <strong>{relic.Description}</strong><br/>
        {separateWithBr(lines)}
      </ReactTooltip>
    );
  }
}
