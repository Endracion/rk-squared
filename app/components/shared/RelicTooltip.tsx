import * as React from 'react';
import * as ReactTooltip from 'react-tooltip';

import { enlir } from '../../data';
import { formatRelicName } from '../../data/items';

import * as _ from 'lodash';

interface Props {
  id: string;
  relicId: number;
}

const attributes = ['atk', 'def', 'mag', 'res', 'mnd', 'acc', 'eva'];

const separateWithBr = (lines: any[]) => {
  const result: any[] = [];
  for (let i = 0; i < lines.length; i++) {
    if (result.length) {
      result.push(<br key={i}/>);
    }
    result.push(lines[i]);
  }
  return result;
};

export class RelicTooltip extends React.PureComponent<Props & any> {
  render() {
    const { id, relicId, ...props } = this.props;
    const relic = enlir.relics[relicId];
    if (!relic) {
      return null;
    }

    const lines = _.filter([
      `${relic.rarity}â˜… ${relic.type}`,
      _.filter(attributes.map(i => relic.stats[i] ? `${i.toUpperCase()} ${relic.stats[i]}` : '')).join(', '),
      relic.effect
    ]);

    return (
      <ReactTooltip id={id} {...props}>
        <strong>{formatRelicName(relic)}</strong><br/>
        {separateWithBr(lines)}
      </ReactTooltip>
    );
  }
}
