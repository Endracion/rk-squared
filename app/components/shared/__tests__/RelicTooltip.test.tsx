import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { RelicTooltip } from '../RelicTooltip';

describe('RelicTooltip', () => {
  it('renders an accessory tooltip', () => {
    const tree = renderer.create(<RelicTooltip id="test" relicId={23080334} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nothing when given an unknown relic', () => {
    const tree = renderer.create(<RelicTooltip id="test" relicId={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
