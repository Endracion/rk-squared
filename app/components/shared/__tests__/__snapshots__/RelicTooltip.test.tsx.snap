// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelicTooltip renders an accessory tooltip 1`] = `
<div
  className="__react_component_tooltip place-top type-dark "
  data-id="tooltip"
  id="test"
>
  <strong>
    Star Pendant (Type-0)
  </strong>
  <br />
  4â˜… Accessory
  <br />
  RES 10
  <br />
  Poison immunity (moderate)
</div>
`;

exports[`RelicTooltip renders nothing when given an unknown relic 1`] = `null`;
